vsr: prevent state sync from breaking hash chain

matklad opened the pull request:
The added test is a minimization of https://github.com/tigerbeetle/tigerbeetle/issues/1532.

If replica state-syncs into a newer view, its log may be disconnected
from a checkpoint. To fix this, require that replica jumps view before
jumping sync target.

The effect here is that, after state sync, there's _still_ a break
between the checkpoint and the log, but, as replica is in a view_change
state, it doesn't assume valid hash chain and waits until an SV to
correct the log.

SEED: 17270152460592452540
Closes: https://github.com/tigerbeetle/tigerbeetle/issues/1532

matklad committed:
vsr: prevent state sync from breaking hash chain

The added test is a minimization of #1532.

If replica state-syncs into a newer view, its log may be disconnected
from a checkpoint. To fix this, require that replica jumps view before
jumping sync target.

The effect here is that, after state sync, there's _still_ a break
between the checkpoint and the log, but, as replica is in a view_change
state, it doesn't assume valid hash chain and waits until an SV to
correct the log.

SEED: 17270152460592452540
Closes: #1532

sentientwaffle commented on src/vsr/replica_test.zig:
```diff
@@ -1138,6 +1139,83 @@ test "Cluster: sync: slightly lagging replica" {
 
+test "Cluster: sync: checkpoint from a newer view" {
+    // B1 appends (but does not commit) prepares across a checkpoint boundary.
+    // Then the cluster truncates those prepares and commits past checkpoint the trigger.
```
Does this mean to say "commits past the checkpoint trigger"?

sentientwaffle commented on src/vsr/replica_test.zig:
```diff
@@ -1138,6 +1139,83 @@ test "Cluster: sync: slightly lagging replica" {
+                        return (prepare.header.op == checkpoint_1);
+                    },
+                    else => return false,
+                }
```
Could be written more succinctly as:

```suggestion
                const prepare = message.into(.prepare) orelse return false;
                return prepare.header.op == checkpoint_1;
```

sentientwaffle commented on src/vsr/replica_test.zig:
```diff
@@ -1138,6 +1139,83 @@ test "Cluster: sync: slightly lagging replica" {
+            }
+        }.drop_message);
+        try c.request(checkpoint_1_trigger + 1, checkpoint_1 - 1);
+        try expectEqual(a0.op_head(), checkpoint_1_trigger - 1);
```
:thinking: I was expecting a0's head to be `checkpoint_1_trigger + 1`. But I suppose with a commit-max of `checkpoint_1 - 1` and both a pipeline and batch-multiple of 4 a0 won't prepare that far ahead. But then why request it?

matklad replied:
>🤔 I was expecting a0's head to be checkpoint_1_trigger + 1.

Huh, I think I expected the same, and ended up with `checkpoint_1_trigger - 1` by accident after fighting unrelated test failures. 

My thinking here is that preparing all the way to past the trigger would be a more interesting test, but, indeed, it seems we can't do that given that pipeline is not longer that batch multiple.

Simplified this to just `checkpoint_1 + 1`, this still triggers the assert.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -8388,6 +8389,17 @@ pub fn ReplicaType(
                 @panic("checkpoint diverged");
             }
 
+            if (candidate.view > self.view) {
+                log.debug("{}: on_{s}: jump_sync_target: ignoring, newer view" ++
+                    " (view={} candidate.view={})", .{
+                    self.replica,
+                    @tagName(header.command),
+                    self.view,
+                    candidate.view,
+                });
+                return;
+            }
```
Checking if I understand how this works:

In `on_start_view()`, we will reject the SV and transition to state sync if all of the SV's headers are beyond our `prepare_max`:

```zig
            for (view_headers.slice) |*header| {
                assert(header.commit <= message.header.commit);

                if (header.op <= self.op_prepare_max()) {
                    if (self.log_view < self.view or
                        (self.log_view == self.view and header.op >= self.op))
                    {
                        self.set_op_and_commit_max(header.op, message.header.commit, @src());
                        assert(self.op == header.op);
                        assert(self.commit_max >= message.header.commit);
                        break;
                    }
                }
            } else {
                // This replica is too far behind, i.e. the new `self.op` is too far ahead of
                // the last checkpoint. If we wrap now, we overwrite un-checkpointed transfers
                // in the WAL, precluding recovery.
                if (self.syncing == .idle) {
                    log.warn("{}: on_start_view: start sync; lagging behind cluster " ++
                        "(op_prepare_max={} quorum_head={})", .{
                        self.replica,
                        self.op_prepare_max(),
                        view_headers.slice[0].op,
                    });
                    self.sync_start_from_committing();
                }
                return;
            }
```

So to advance our view (such that this new `jump_sync_target()` will not drop the message) we rely on pings or other messages that `jump_view()` uses to transition to view-change status (_not_ normal status, since we will ignore the SV that we would request).

And `jump_view()` does indeed precede `jump_sync_target()`, so that can happen with a single ping:

```zig
            self.jump_view(message.header);
            self.jump_sync_target(message.header);
```

matklad replied:
Not exactly! This is related to me saying "the code already has the fix". My _original_ thinking was that, indeed, accepting SV requires to do the state sync first. But to do state sync we need to change view. And when we receive a commit we don't jump view immediately, but wait for SV, which seems to create a deadlock.

So the fix I was planing to implement here was to jump to `view_change` when we receive a checkpoint target from the next log wrap _and_ the next view. 

But, when I started coding, I realised that there's already code to do that! Right here:

https://github.com/tigerbeetle/tigerbeetle/blob/7daa0b6098c6a066ff3dd4d68ee2f35b55da6599/src/vsr/replica.zig#L1933-L1935

That is, even though we don't append any headers from an SV, we still go into view_change! And that resolves the deadlock. 

So the full sequence of events is this:

* we receive a  `.commit` with the next view & checkpoint
* when `jump_view` on this commit, we send out `.request_start_view`
* when `jump_sync_target` on this commit, we ignore the target, as it's in the future view
* sometime later, we receive the requested SV
* if we install anything from that SV, we transition to .normal in the new view and can proceed with state sync
* if we don't install anything, we trantisition to .view_change in the new view, and can still proceed with state sync (and will ask for another SV after sync is done)

sentientwaffle replied:
Ahh that makes sense, I forgot that `on_start_view` did that!

matklad replied:
This made merealize that there's a subtlety here! I forgot about `view_durable` and check only for `view`, but this is still correct:

when view_durable is not updated, we _can_ crash and restart into an earlier view, but that also means that we'll restart _without_ newer sync target, so it ends up ok in the end. 

Added a maybe to that effect.

matklad committed:
vsr: clarify that checking view_durable is not needed in jump_sync

It is correct to check only view, not view_durable here. This can't lead
to a situation where we crash and restart with an older view and a newer
sync target, because superblock updates are serialized.
